libcst/matchers/_matcher_base.py

```diff
@@ -446,7 +472,93 @@ def __invert__(self) -> "MatchMetadata":
         return cast(MatchMetadata, InverseOf(self))
 
     def __repr__(self) -> str:
-        return f"MatchMetadata(key={repr(self.key)}, value={repr(self.value)})"
+        return f"MatchMetadata(key={repr(self._key)}, value={repr(self._value)})"
+
+
+class MatchMetadataIfTrue(_BaseMetadataMatcher):
+    """
+    Matcher that looks up the metadata on the current node using the provided
+    metadata provider and passes it to a callable which can inspect the metadata
+    further, returning ``True`` if the matcher should be considered a match.
+    If the metadata value does not exist for a particular node,
+    :class:`MatchMetadataIfTrue` will always be considered not a match.
+
+    For example, to match against any arg whose qualified name might be
+    ``typing.Dict``::
+
+        m.Call(
+            args=[
+                m.Arg(
+                    m.MatchMetadataIfTrue(
+                        meta.QualifiedNameProvider,
+                        lambda qualnames: any(n.name == "typing.Dict" for n in qualnames)
+                    )
+                )
+            ]
+        )
+
+    To match against any :class:`~libcst.Name` node for the identifier ``foo``
+    as long as that identifier is found at the beginning of an unindented line::
+
+        m.Name(
+            value="foo",
+            metadata=m.MatchMetadataIfTrue(
+                meta.SyntacticPositionProvider,
+                lambda position: position.start.column == 0,
+            )
+        )
+
+    This can be used in place of any concrete matcher as long as it is not the
+    root matcher. Calling :func:`matches` directly on a :class:`MatchMetadataIfTrue`
```

Comment:
What happens when it's used as a root matcher?

Reply:
It does still work in the base case, but I've intentionally narrowed the types on `matches` so that it is a mypy/pyre error. For @visit and @leave nodes, it will error out with the runtime type checking since you can't infer types.